builder/dockerfile/typedcommand/parse.go

```diff
@@ -0,0 +1,638 @@
+package typedcommand
+
+import (
+	"fmt"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/strslice"
+	"github.com/docker/docker/builder/dockerfile/command"
+	"github.com/docker/docker/builder/dockerfile/parser"
+	"github.com/docker/go-connections/nat"
+	metrics "github.com/docker/go-metrics"
+	"github.com/pkg/errors"
+)
+
+const (
+	metricsUnknownInstructionError = "unknown_instruction_error"
+)
+
+type parseRequest struct {
+	args       []string
+	attributes map[string]bool
+	flags      *BFlags
+	original   string
+}
+
+func nodeArgs(node *parser.Node) []string {
+	result := []string{}
+	for ; node.Next != nil; node = node.Next {
+		arg := node.Next
+		result = append(result, arg.Value)
+	}
+	return result
+}
+
+func newParseRequestFromNode(node *parser.Node) parseRequest {
+	return parseRequest{
+		args:       nodeArgs(node),
+		attributes: node.Attributes,
+		original:   node.Original,
+		flags:      NewBFlagsWithArgs(node.Flags),
+	}
+}
+
+func ParseCommand(node *parser.Node, buildsFailed metrics.LabeledCounter) (interface{}, error) {
+	switch node.Value {
+	case command.Env:
+		return parseEnv(newParseRequestFromNode(node))
+	case command.Maintainer:
+		return parseMaintainer(newParseRequestFromNode(node))
+	case command.Label:
+		return parseLabel(newParseRequestFromNode(node))
+	case command.Add:
+		return parseAdd(newParseRequestFromNode(node))
+	case command.Copy:
+		return parseCopy(newParseRequestFromNode(node))
+	case command.From:
+		return parseFrom(newParseRequestFromNode(node))
+	case command.Onbuild:
+		return parseOnbuild(newParseRequestFromNode(node))
+	case command.Workdir:
+		return parseWorkdir(newParseRequestFromNode(node))
+	case command.Run:
+		return parseRun(newParseRequestFromNode(node))
+	case command.Cmd:
+		return parseCmd(newParseRequestFromNode(node))
+	case command.Healthcheck:
+		return parseHealthcheck(newParseRequestFromNode(node))
+	case command.Entrypoint:
+		return parseEntrypoint(newParseRequestFromNode(node))
+	case command.Expose:
+		return parseExpose(newParseRequestFromNode(node))
+	case command.User:
+		return parseUser(newParseRequestFromNode(node))
+	case command.Volume:
+		return parseVolume(newParseRequestFromNode(node))
+	case command.StopSignal:
+		return parseStopSignal(newParseRequestFromNode(node))
+	case command.Arg:
+		return parseArg(newParseRequestFromNode(node))
+	case command.Shell:
+		return parseShell(newParseRequestFromNode(node))
+	}
+
+	buildsFailed.WithValues(metricsUnknownInstructionError).Inc()
+	return nil, fmt.Errorf("unknown instruction: %s", strings.ToUpper(node.Value))
+}
+
+func Parse(ast *parser.Node, targetStage string, buildsFailed metrics.LabeledCounter) (stages BuildableStages, metaArgs []ArgCommand, err error) {
```

Comment:
BuildableStages is a type alias for []BuildableStage with helper methods. Do you prefer explicit []BuildableStage return type and convertion of methods to free functions ?